src/engines/basic-render-engine.js

```diff
@@ -159,69 +157,70 @@ export class BasicRenderEngine extends EventEmitter {
     }
 
     addRectToRenderQueue(color, x, y, w) {
-        if (!this.rectRenderQueue[color]) {
-            this.rectRenderQueue[color] = [];
-        }
-
-        this.rectRenderQueue[color].push({ x, y, w });
+        this.renderQueue.push({ type: 'rect', value: { color, x, y, w } });
     }
 
     addTextToRenderQueue(text, x, y, w) {
         if (text) {
             const textMaxWidth = w - (this.blockPaddingLeftRight * 2 - (x < 0 ? x : 0));
 
             if (textMaxWidth > 0) {
-                this.textRenderQueue.push({ text, x, y, w, textMaxWidth });
+                this.renderQueue.push({ type: 'text', value: { text, x, y, w, textMaxWidth } });
             }
         }
     }
 
     addStrokeToRenderQueue(color, x, y, w, h) {
-        this.strokeRenderQueue.push({ color, x, y, w, h });
-    }
-
-    resolveRectRenderQueue() {
-        Object.entries(this.rectRenderQueue).forEach(([color, items]) => {
-            this.setCtxColor(color);
-
-            items.forEach(({ x, y, w }) => this.renderBlock(color, x, y, w));
-        });
-
-        this.rectRenderQueue = {};
-    }
-
-    resolveTextRenderQueue() {
-        this.setCtxColor(this.styles.fontColor);
-
-        this.textRenderQueue.forEach(({ text, x, y, w, textMaxWidth }) => {
-            const { width: textWidth } = this.ctx.measureText(text);
-
-            if (textWidth > textMaxWidth) {
-                const avgCharWidth = textWidth / (text.length);
-                const maxChars = Math.floor((textMaxWidth - this.placeholderWidth) / avgCharWidth);
-                const halfChars = (maxChars - 1) / 2;
-
-                if (halfChars > 0) {
-                    text = text.slice(0, Math.ceil(halfChars)) + '…' + text.slice(text.length - Math.floor(halfChars), text.length);
-                } else {
-                    text = '';
-                }
-            }
-
-            if (text) {
-                this.ctx.fillText(text, (x < 0 ? 0 : x) + this.blockPaddingLeftRight, y + this.blockHeight - this.blockPaddingTopBottom);
+        this.renderQueue.push({ type: 'stroke', value: { color, x, y, w, h } });
+    }
+
+    resolveRenderQueue() {
+        this.renderQueue.forEach(({type, value}) => {
+            switch (type) {
```

Comment:
I fell like this section could be refactored.... its a straight up shift of the individual renders for the most part and could probably be moved into their own `resolveRect`, `resolveStroke`, `resolveTest` functions.